io/io/src/TBufferFile.cxx

```diff
@@ -2858,21 +3213,49 @@ Version_t TBufferFile::ReadVersion(UInt_t *startpos, UInt_t *bcnt, const TClass
       Version_t  vers[2];
    } v;
 #ifdef R__BYTESWAP
-   frombuf(this->fBufCur,&v.vers[1]);
-   frombuf(this->fBufCur,&v.vers[0]);
+   if (IsBufBigEndian()) {
+      frombuf(this->fBufCur,&v.vers[1]);
+      frombuf(this->fBufCur,&v.vers[0]);
+   } else {
+      frombuf(this->fBufCur,&v.vers[0],0);
+      frombuf(this->fBufCur,&v.vers[1],0);
+   }
 #else
-   frombuf(this->fBufCur,&v.vers[0]);
-   frombuf(this->fBufCur,&v.vers[1]);
+   if (IsBufBigEndian()) {
+      frombuf(this->fBufCur,&v.vers[0]);
+      frombuf(this->fBufCur,&v.vers[1]);
+   } else {
+      frombuf(this->fBufCur,&v.vers[1],0);
+      frombuf(this->fBufCur,&v.vers[0],0);
+   }
 #endif
-
-   // no bytecount, backup and read version
-   if (!(v.cnt & kByteCountMask)) {
-      fBufCur -= sizeof(UInt_t);
-      v.cnt = 0;
+   // no bytecount, backup and read version.
+   // read in next UShort_t and determine if it is zero or not. 
+   // If it's not zero, it is 'highly possible' storing version
+   // and consequently the first four byte should be byte count.
```

Comment:
Does the comment applies to both side if the next if statement?  The code is quite different between the two case, and it deserve some explanation why.